Add pipeline support

XiongKezhi opened the pull request:
Target pipeline:
```
publishChecks(
    name: "check", 
    detailsURL: "ci.jenkins.io", 
    status: "COMPLETED",
    conclusion: "SUCCESS",
    output: [
        title: "output",
        summary: "summary",
        text: "text",
        annotations: [
            [path: "src/main/foo.java", startLine: 0, endLine: 0, annotationLevel: "WARNING", message: "foo test"],
            [path: "src/main/bar.java", startLine: 1, endLine: 1, annotationLevel: "WARNING", message: "bar test"]
        ],
        images: [
            [alt: "foo alt", imageUrl: "foo url", caption: "foo image"],    
            [alt: "foo alt", imageUrl: "foo url", caption: "foo image"]
        ]
    ],
    actions: [
        [label: "foo label", description: "foo action", identifier: "foo identifier"],
        [label: "bar label", description: "bar action", identifier: "bar identifier"]
    ]
)

```

- [x] Implement publish checks step
- [x] Add tests

timja commented on src/main/java/io/jenkins/plugins/checks/api/ChecksAnnotation.java:
```diff
@@ -12,16 +17,41 @@
-    private final String rawDetails;
+@SuppressFBWarnings(value = "NP_NONNULL_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR",
+        justification = "Empty constructor used by stapler")
+public class ChecksAnnotation extends AbstractDescribableImpl<ChecksAnnotation> implements Serializable {
```
Is it worth separating our pipeline models from our API models?

So that they don't get polluted with all the setters etc required by stapler?

timja commented on src/main/java/io/jenkins/plugins/checks/steps/PublishChecksStep.java:
```diff
@@ -0,0 +1,142 @@
+    private ChecksConclusion conclusion;
+    private LocalDateTime startedAt = LocalDateTime.now();
+    private LocalDateTime completedAt = LocalDateTime.now();
+    private ChecksOutput output;
```
I really think we should flatten this down so that users don't have to do:

```
publishChecks(
    name: "check", 
    detailsURL: "ci.jenkins.io", 
    status: "COMPLETED",
    conclusion: "SUCCESS",
    output: [
        title: "output",
        summary: "summary",
        text: "text",
        annotations: [
            [path: "src/main/foo.java", startLine: 0, endLine: 0, annotationLevel: "WARNING", message: "foo test"],
            [path: "src/main/bar.java", startLine: 1, endLine: 1, annotationLevel: "WARNING", message: "bar test"]
        ],
        images: [
            [alt: "foo alt", imageUrl: "foo url", caption: "foo image"],    
            [alt: "foo alt", imageUrl: "foo url", caption: "foo image"]
        ]
    ],
)
```


instead a simple usage would be;

```
publishChecks(
  name: "MyCheck",
  conclusion: "success",
  summary: "Everything is OK!"
)
```

i.e. it would be good to start from a users perspective, and look at some scenarios of what they want to use it for.

Most users probably won't user annotations / images
They will just want to set a status, with some messaging I would expect

XiongKezhi replied:
yeah, I didn't consider the user scenarios and afraid too many fields will explode this class, so just flattened the first level checks `ChecksDetails`. 

After your example, one pipeline usage I can come up with now is that users may want to publish checks to more precisely indicate the stages (especially for their customized ones).

XiongKezhi commented:
For now, I think it's ok we do not need to provide annotations and images, but I think it would be better to provide a new `action` object for the pipeline step: after we support the re-run request, users can create re-run action in pipeline.  What do you think? @timja

timja commented:
Hmm not sure if actions is too useful it sends a webhook event to the app, and in order to do anything with it you need a Jenkins plugin...

XiongKezhi commented:
> Hmm not sure if actions is too useful it sends a webhook event to the app, and in order to do anything with it you need a Jenkins plugin...
> 
> 
> 
> 

But we'll need to handle the webhook event for the re-run request anyway, there is a identifier for each kind of action, so we can tell the users to use the re-run identifier if they want to add the action.

timja commented:
> > Hmm not sure if actions is too useful it sends a webhook event to the app, and in order to do anything with it you need a Jenkins plugin...
> 
> But we'll need to handle the webhook event for the re-run request anyway, there is a identifier for each kind of action, so we can tell the users to use the re-run identifier if they want to add the action.

if there's nothing we can do about the webhook event we'd just have to ignore it.

We could have another step which is wait for github checks event, where you pass the identifier and we attach a pauseaction to the build...

but that's more than a v1, I would add than in a follow up, if we could get an MVP first that would be great

uhafner commented on src/main/java/io/jenkins/plugins/checks/steps/PublishChecksStep.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    @DataBoundSetter
+    public void setDetailsURL(final String detailsURL) {
```
Shouldn't this be `detailsUrl`? (If not make sure that `imageUrl` will use the same style)

ianfixes commented:
Joining the discussion here [via Gitter](https://gitter.im/jenkinsci/github-checks-api?at=5f3146e19b76045d5b8d6a8b), I'll briefly summarize some of the stuff I ran into since posting on [JENKINS-51225](https://issues.jenkins-ci.org/browse/JENKINS-51225?focusedCommentId=381991&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-381991).

1. Jenkins can run tests both on the source and target branch of a pull request (i.e. the branch by itself, and then the result of merging the source branch with the target branch.  This means that the value you have for `GIT_COMMIT` might not be one that GitHub has ever heard of.  We hack around this by running `git log --pretty='%H %h %ae' -n 3` and picking the first commit that isn't authored by `nobody@nowhere`.
2. Assuming that you've now found the "true" commit of a merged PR branch, you are now in the uncomfortable position of getting _actual_ static analysis errors but not being sure whether the line numbers have shifted due to the merge!  So in this case, we still report the overall conclusion of the test (pass, fail, neutral) but summarize the results in a table instead of annotating what might be the wrong line.
3. Now that you're annotating the source branch and the merged branch in parallel, note that GitHub will refuse to report 2 different sets of annotations with the same name -- so you'll want to adjust the report name somewhat.  (We append ` (PR)` to the report name if it's a pull request branch).  
4. Now that you're reporting 2 sets of results properly, you may notice that there can be thousands of annotations.  GitHub's API wants you to split these into batches of 50.  Instead of reporting every possible batch of annotations, we've found that having 50 errors is more than enough for an engineer to work on -- we only report the first 50.  But to do that intelligently, we need to make sure that the most severe errors are promoted to the top of that list.
5. Now that you're reporting the most severe errors, you might notice that _you might be reporting on files that were **not** actually part of the pull request!_.  So in the case of a PR test, we have the advantage of being able to tell what files had changed between source and target branch: we use a tip from [JENKINS-54285](https://issues.jenkins-ci.org/browse/JENKINS-54285?focusedCommentId=354715#comment-354715) with the addition of `git fetch origin ${env.CHANGE_TARGET}:refs/remotes/origin/${env.CHANGE_TARGET}` to ensure that we have the proper local data.  This gives us the list of files that have actually changed, and we prioritize annotations in this change set before all others.
6. Finally, we have derived incredible value from _overriding the severity_ of reports that we get from other tools.  For example, we may be trying out a new static analysis tool but figuring out what settings to use.  Or we may want to report errors as warnings to allow a transition period to a new set of engineering best practices. Or we may even want to try out new, stricter settings for a tool we already use.  

Our implementation of this is centered around uploading annotations in the structure that GitHub expects; this general-purpose function enforces the naming (PR or not PR) and prioritization of annotations based on source file and severity.  We then have a supplementary set of functions to translate from common report formats (eslint, rubocop, etc) into GitHub format -- each of which accepts a closure for adjusting the severity of an annotation entry.  With that, we have everything we need to support a lot of interesting use cases.

So with that said, I really like the flexibility being proposed here:

```groovy
publishChecks(
    name: "check", 
    detailsURL: "ci.jenkins.io", 
    status: "COMPLETED",
    conclusion: "SUCCESS",
    output: [
        title: "output",
        summary: "summary",
        text: "text",
        annotations: [
            [path: "src/main/foo.java", startLine: 0, endLine: 0, annotationLevel: "WARNING", message: "foo test"],
            [path: "src/main/bar.java", startLine: 1, endLine: 1, annotationLevel: "WARNING", message: "bar test"]
        ],
        images: [
            [alt: "foo alt", imageUrl: "foo url", caption: "foo image"],    
            [alt: "foo alt", imageUrl: "foo url", caption: "foo image"]
        ]
    ],
)
```

Although the `detailsURL` and `status` should not be up to the user to supply -- Jenkins knows what those values are.  Also, I would break up `output` into the 3 things that are genuinely up to the user: title, description, and a set of annotations.  The summary (as indicated above) may need to be constructed from the annotations in the case of a pull request branch.

Sorry this was long, and I accept in advance that my use cases very likely don't represent "most users" 😄 

Ask me anything about how this is working for our team so far!

XiongKezhi commented:
@ianfixes Thanks for sharing the issues you met on your implementation and feedback very much! 👏

Many of your problems we have met as well, here are some of our solutions we implemented:

For the true commit of PR, our implementation is based on [GitHub Branch Source Plugin](https://github.com/jenkinsci/github-branch-source-plugin) which is an implementation of the [Branch API](https://github.com/jenkinsci/branch-api-plugin), so resolving the true commit of PR isn't a hard problem after that. 

For the limit of annotations, our depending library did that for us: https://github.com/hub4j/github-api/blob/master/src/main/java/org/kohsuke/github/GHCheckRunBuilder.java#L138

For test and static warnings report, we have implemented the consumers in other plugins, you can find them in our blog post: https://www.jenkins.io/blog/2020/08/03/github-checks-api-plugin-coding-phase-2/. We only reported the new issues of the PR comparing with the target branch, so resolving the problem about "files that were not actually part of the pull request".

ianfixes commented:
Thanks XiongKezhi!  I figured that you would have run into a few of these and resolved them.

After thinking about it, I like the idea of this syntax:

```groovy
// specify everything manually
publishChecks(
    name: "my check", 
    conclusion: "SUCCESS",
    output: [
        title: "output",
        summary: "summary",
        text: "text",
        annotations: [
            [path: "src/main/foo.java", startLine: 0, endLine: 0, annotationLevel: "WARNING", message: "foo test"],
            [path: "src/main/bar.java", startLine: 1, endLine: 1, annotationLevel: "WARNING", message: "bar test"]
        ],
        images: [
            [alt: "foo alt", imageUrl: "foo url", caption: "foo image"],    
            [alt: "foo alt", imageUrl: "foo url", caption: "foo image"]
        ]
    ]
)
```

Or for using a built-in converter/tool to read a report from a different format (if I'm reading [the page you referred to](https://www.jenkins.io/blog/2020/08/03/github-checks-api-plugin-coding-phase-2/#try-it) correctly): 

```groovy
// "formatIssues" style borrowed from "recordIssues"
publishChecks(
    name: "my check", 
    conclusion: "SUCCESS",
    output: formatIssues(tools: [checkStyle(pattern: 'target/checkstyle-result.xml'),
                         spotBugs(pattern: 'target/spotbugsXml.xml'),
                         pmdParser(pattern: 'target/pmd.xml'),
                         cpd(pattern: 'target/cpd.xml')],
                         qualityGates: [[threshold: 1, type: 'TOTAL', unstable: true]])
)


```

XiongKezhi commented:
@ianfixes  Is there any reason that you don't want to use warnings-ng to help publish tool reports but want to convert them by hand?

XiongKezhi commented:
Strange, the PR passes my local tests while the CI says
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project checks-api: Compilation failure
[ERROR] /home/runner/work/checks-api-plugin/checks-api-plugin/src/main/java/io/jenkins/plugins/checks/api/ChecksImage.java:[79,36] cannot find symbol
[ERROR]   symbol:   variable imageUrl
[ERROR]   location: class io.jenkins.plugins.checks.api.ChecksImage
```

And there is no line#79 at all in `ChecksImage.java`

ianfixes commented:
> Is there any reason that you don't want to use warnings-ng to help publish tool reports

Perhaps the biggest reason is that we've found working with plugins to be terrible, especially for scripted pipeline.  And in many cases, the functionality is trivial enough for us to just implement ourselves in a script library such that we have full control over it.

In the case of publishing reports specifically, as I mentioned above sometimes we want to adjust the severity of the annotations that a tool provides.  For example, we may want to soft-launch some style guidelines such that the errors are reduced to warnings for some grace period.  Or we want to test out a proposed linting tool, in much the same way. Or we want an ability to introduce our own logic to mute certain errors.  

Our feeling is that the tools are there to help us, and not the other way around.  So we leave room to tailor the reports to our own needs.

uhafner commented on src/test/java/io/jenkins/plugins/checks/steps/PublishChecksStepTest.java:
```diff
@@ -0,0 +1,86 @@
+        PublishChecksStep.PublishChecksStepDescriptor descriptor = new PublishChecksStep.PublishChecksStepDescriptor();
+        assertThat(descriptor.getFunctionName()).isEqualTo("publishChecks");
+        assertThat(descriptor.getDisplayName()).isEqualTo("Publish customized checks to SCM platforms");
+        assertThat(descriptor.getRequiredContext())
+                .containsAll(new HashSet(Arrays.asList(Run.class, TaskListener.class)));
```
```suggestion
        assertThat(descriptor.getRequiredContext().toArray())
                .contains(Run.class, TaskListener.class);
```

uhafner reviewed (changes requested):
I think it would help if we add the help texts (and the Snippet Generator configuration). 

I think it would be also helpful to have a real integration test that creates a sample pipeline using the step. Then we can see in the console log if the null publisher is correctly invoked in the API plugin. 
Examples: https://github.com/jenkinsci/warnings-ng-plugin/blob/master/plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java
(If this is too much work for now we can also add such a test to the GitHub checks plugin).

uhafner commented on src/main/java/io/jenkins/plugins/checks/steps/PublishChecksStep.java:
```diff
@@ -0,0 +1,197 @@
+        public ListBoxModel doFillStatusItems() {
+            ListBoxModel options = new ListBoxModel();
+            for (ChecksStatus status : ChecksStatus.values()) {
+                options.add(StringUtils.capitalize(status.name().toLowerCase(Locale.ENGLISH).replace("_", " ")),
```
Extract to a method `asDisplayName` or something similar and reuse in line 154.

uhafner replied:
(Or extract the whole loop as method that accepts any enum)

uhafner commented on src/main/java/io/jenkins/plugins/checks/steps/PublishChecksStep.java:
```diff
@@ -0,0 +1,140 @@
+        @NonNull
+        @Override
+        public String getDisplayName() { // it's pipeline step, so where does the name shown?
+            return "Publish customized checks to SCM platforms";
```
Optional (since we have no I18n yet in the whole project): it would make sense to create a `Messages.properties` file and add all UI labels (example: https://github.com/jenkinsci/forensics-api-plugin/blob/master/src/main/resources/io/jenkins/plugins/forensics/miner/Messages.properties). Then you can replace these strings with the constant in the generated `Messages` class (example: https://github.com/jenkinsci/forensics-api-plugin/blob/master/src/main/java/io/jenkins/plugins/forensics/miner/ForensicsJobAction.java#L34). 

But we can postpone that part to the after GSoC work.

uhafner commented on src/main/resources/io/jenkins/plugins/checks/steps/PublishChecksStep/conclusion.html:
```diff
@@ -0,0 +1,5 @@
+<div>
```
Shouldn't this file be named `help-conclusion.html`?

uhafner commented on src/main/resources/io/jenkins/plugins/checks/steps/PublishChecksStep/status.html:
```diff
@@ -0,0 +1,3 @@
+<div>
```
Shouldn't this file be named `help-status`?

uhafner reviewed (approved):
Looks good to me! Just some minor things.
